app/models/conversion_host/configurations.rb

```diff
@@ -0,0 +1,78 @@
+module ConversionHost::Configurations
+  extend ActiveSupport::Concern
+  class ConfigurationError < StandardError
+  end
+  module ClassMethods
+    def notify_configuration_result(op, success, resource_info)
+      Notification.create(
+        :type    => success ? :conversion_host_config_success : :conversion_host_config_failure,
+        :options => {
+          :op_name => op,
+          :op_arg  => resource_info,
+        }
+      )
+    end
+
+    def queue_configuration(op, instance_id, resource, params, auth_user = nil)
+      task_opts = {
+        :action => "Configuring a conversion_host: operation=#{op} resource=(type: #{resource.class.name} id:#{resource.id})",
+        :userid => auth_user
+      }
+      queue_opts = {
+        :class_name  => name,
+        :method_name => op,
+        :instance_id => instance_id,
+        :role        => 'ems_operations',
+        :zone        => resource.ext_management_system.my_zone,
+        :args        => params
+      }
+      MiqTask.generic_action_with_callback(task_opts, queue_opts)
+    end
+
+    def enable_queue(params, auth_user = nil)
+      resource_type = params[:resource_type]
+      resource_id = params[:resource_id]
+      resource = resource_type.constantize.find(resource_id)
+
+      queue_configuration('enable', nil, resource, [params], auth_user)
+    end
+
+    def enable(params)
+      _log.info("Enabling a conversion_host with parameters: #{params}")
+      params.delete(:task_id) # in case this is being called through *_queue which will stick in a :task_id
+      params.delete(:miq_task_id) # miq_queue.activate_miq_task will stick in a :miq_task_id
+
+      vddk_url = params.delete("param_v2v_vddk_package_url")
+      resource_id = params[:resource_id]
+      resource_type = params[:resource_type]
+      resource = resource_type.constantize.find(resource_id)
+
+      conversion_host = new(params.merge(:resource => resource))
+      conversion_host.enable_conversion_host_role(vddk_url)
+      conversion_host.save!
+      conversion_host
+    rescue => error
+      raise(ConfigurationError, error.message)
+    ensure
+      resource_info = "type=#{params[:resource_type]} id=#{params[:resource_id]}"
+      notify_configuration_result('enable', error.nil?, resource_info)
+    end
+
+    def disable_queue(id, auth_user = nil)
+      conversion_host = find(id)
+      queue_configuration('disable', id, conversion_host.resource, [], auth_user)
+    end
+  end
+
+  def disable
+    resource_info = "type=#{resource.class.name} id=#{resource.id}"
+    _log.info("Disabling a conversion_host #{resource_info}")
+
+    disable_conversion_host_role
+    destroy!
+  rescue => error
+    raise(ConfigurationError, error.message)
```

Comment:
Do you really need a new error class for this?  Usually you'd do this to differentiate between "system" errors and errors that you raise and only catch the latter in your callers, but to catch StandardError and re-raise ConfigurationError...might as well just let the original exception get raised.

Reply:
Wondered about that myself, thought maybe there was some sort of `Configurations` interface that needed to be matched or something.

Reply:
I think my problem is that I don't know what are the errors the actual operation could raise.
Maybe I'll just re-raise it.

Reply:
Question, do we actually want to raise errors within these methods, or just log it and issue a warning? And perhaps that warning could be picked up by the UI?

Reply:
No need to catch and re-raise, MiqQueue#deliver will handle logging the exception for you

Reply:
The `ensure` block will send a notification in both case of success/failure

Reply:
The catch is allow the notification to be aware of error situation

Reply:
Oh okay, i'm good with the way you have it now then (`rescue => error; raise; ensure`)